test/unit/postprocessing/test_threshold_optimization.py

```diff
@@ -320,144 +314,123 @@ def _average_prediction_for_label(label, sensitive_feature_value,
+        _POS_P1: 0.66733333,
+        _NEG_P0: 0.666,
+        _NEG_P1: 0.334
+    },
+    "example 2": {
+        _POS_P0: 0.112,
+        _POS_P1: 0.888,
+        _NEG_P0: 0.334,
+        _NEG_P1: 0.666
+    },
+    "example 3": {
+        _POS_P0: 0.33333333333333337,
+        _POS_P1: 0.6666666666666666,
+        _NEG_P0: 0.5,
+        _NEG_P1: 0.5
+    }
+}
+
+
+def test_threshold_optimization_equalized_odds_e2e(data_X_y_sf):
     adjusted_predictor = ThresholdOptimizer(unconstrained_predictor=ExamplePredictor(),
                                             constraints=EQUALIZED_ODDS)
-    adjusted_predictor.fit(X, y, sensitive_features=sensitive_features_)
+    adjusted_predictor.fit(data_X_y_sf.X, data_X_y_sf.y,
+                           sensitive_features=data_X_y_sf.sensitive_features)
+
+    predictions = adjusted_predictor._pmf_predict(
+        data_X_y_sf.X, sensitive_features=data_X_y_sf.sensitive_features)
 
-    predictions = adjusted_predictor._pmf_predict(X, sensitive_features=sensitive_features_)
+    expected_ps = _expected_ps_equalized_odds[data_X_y_sf.example_name]
+    mapped_sensitive_features = _map_into_single_column(data_X_y_sf.sensitive_features)
 
     # assert equalized odds
-    for a in sensitive_feature_names:
-        positive_indices = (np.array(sensitive_features) == a) * \
-            (np.array(labels_ex) == 1)
-        negative_indices = (np.array(sensitive_features) == a) * \
-            (np.array(labels_ex) == 0)
-        average_probs_positive_indices = np.average(
-            predictions[positive_indices], axis=0)
-        average_probs_negative_indices = np.average(
-            predictions[negative_indices], axis=0)
-        assert np.isclose(
-            average_probs_positive_indices[0], expected_positive_p0)
-        assert np.isclose(
-            average_probs_positive_indices[1], expected_positive_p1)
-        assert np.isclose(
-            average_probs_negative_indices[0], expected_negative_p0)
-        assert np.isclose(
-            average_probs_negative_indices[1], expected_negative_p1)
-
-
-@pytest.mark.parametrize("sensitive_features,sensitive_feature_names",
-                         [(sensitive_features_ex1, sensitive_feature_names_ex1),
-                          (sensitive_features_ex2, sensitive_feature_names_ex2)])
-@pytest.mark.parametrize("X_transform", ALLOWED_INPUT_DATA_TYPES)
-@pytest.mark.parametrize("y_transform", ALLOWED_INPUT_DATA_TYPES)
-@pytest.mark.parametrize("sensitive_features_transform", ALLOWED_INPUT_DATA_TYPES)
+    for a in data_X_y_sf.feature_names:
+        pos_indices = (mapped_sensitive_features == a) * (labels_ex == 1)
+        neg_indices = (mapped_sensitive_features == a) * (labels_ex == 0)
+        average_probs_positive_indices = np.average(predictions[pos_indices], axis=0)
+        average_probs_negative_indices = np.average(predictions[neg_indices], axis=0)
+        assert np.isclose(average_probs_positive_indices[0], expected_ps[_POS_P0])
```

Comment:
What are the differences between `np.isclose` and `pytest.approx.` ?

Reply:
> 
> 
> I am a little concerned that this is a monster PR going in lots of different directions at once. Are there any pieces which can be split off relatively easily, or is it too tangled together?

I keep going back to this but 90% of the code changes are in tests. I cold apply the improved input validation only to postprocessing, or only to reductions, and then only adjust their tests, but it mainly saves on changed test lines, not in actual logic.

Reply:
isclose: https://github.com/numpy/numpy/blob/v1.17.0/numpy/core/numeric.py#L2179-L2292
approx: https://docs.pytest.org/en/latest/_modules/_pytest/python_api.html#approx

The second link says this:
```

    If you're thinking about using ``approx``, then you might want to know how
    it compares to other good ways of comparing floating-point numbers.  All of
    these algorithms are based on relative and absolute tolerances and should
    agree for the most part, but they do have meaningful differences:

...

    - ``numpy.isclose(a, b, rtol=1e-5, atol=1e-8)``: True if the difference
      between ``a`` and ``b`` is less that the sum of the relative tolerance
      w.r.t. ``b`` and the absolute tolerance.  Because the relative tolerance
      is only calculated w.r.t. ``b``, this test is asymmetric and you can
      think of ``b`` as the reference value.  Support for comparing sequences
      is provided by ``numpy.allclose``.  `More information...`__

Reply:
I don't see this as a major issue. We can make it the same throughout our tests at some point if we feel like it